sys/kern/vm_map.c

```diff
@@ -164,25 +164,50 @@ static vm_map_entry_t *vm_map_entry_split(vm_map_t *map, vm_map_entry_t *ent,
   return new_ent;
 }
 
-void vm_map_entry_destroy_range(vm_map_t *map, vm_map_entry_t *ent,
-                                vaddr_t start, vaddr_t end) {
+static int vm_map_destroy_range_nolock(vm_map_t *map, vaddr_t start,
+                                       vaddr_t end) {
   assert(mtx_owned(&map->mtx));
-  assert(start >= ent->start && end <= ent->end);
+
+  /* Find first entry affected by unmapping memory. */
+  vm_map_entry_t *ent = vm_map_find_entry(map, start);
+  if (!ent)
+    return 0;
 
   pmap_remove(map->pmap, start, end);
```

Comment:
Please remember that page tables are reconstructed lazily based on `vm_map`, i.e. `vm_map` is the only source of truth when it comes to shape of the address space.

In a sense `pmap` caches recently used parts of `vm_map`. After performing operation on `vm_map` you must take care of removing possibly invalid entries in "the cache". Hence you should make a call to `pmap_remove` after you modify `vm_map`.

Reply:
I have found that some operation expect that `pmap_remove` is called before them (e.g. removing `vm_object`). 

I also think that we can remove entries from page table here, because we are under `vm_map` lock. It ensures that no one will modify the `vm_map` and `pmap` when we are removing `vm_map_entries`. New mappings can be added only in `vm_page_fault` which also requires the `vm_map` lock.